Refactor ZTF matchfile ingestion

kmshin1397 opened the pull request:
Refactor the `ingest_ztf_matchfiles.py` script to be more performant. 
- Now reads the entire `sources` and `sourcedata` table in a matchfile at once into Pandas DataFrames, which are then joined on `matchid` instead of looping through the sources and fetching the corresponding `sourcedata` for each. 
- Type conversions of Numpy subtypes for MongoDB insertion is done without explicit type-checking with `np.issubdtype()`. These calls were a significant portion of the computation time. Instead, we just record the columns we expect to be `ints` and assume the others are `floats` so that we can just do constant-time type checking of fields against the set of int fields. 
- Not performance-related, but added a custom `istarmap` function to `multiprocessing` to get more responsive `tqdm` progress bars for the multiprocessing worker pool (the previous version waited for all subprocesses to finish before updating the progress, i.e. 0% to 100% only) https://stackoverflow.com/questions/57354700/starmap-combined-with-tqdm
- Removed the `keep_all` option and just filter out all excluded columns always.

kmshin1397 commented:
Addresses #72

dmitryduev commented on tools/ingest_ztf_matchfiles.py:
```diff
@@ -209,186 +285,89 @@ def process_file(
+                np.seterr(**old_settings)
+                sources["iqr"] = iqr
+
+                sourcedatas = pd.DataFrame.from_records(
```
Would you mind measuring how much processing time it would add if you were to sort by "hjd" (which is a time stamp - heliocentric Julian date)? I've noticed that there hasn't been any sorting of the data field. No one has ever complained, but I think it would be more natural to store a sorted array. If it adds more than 5%, let's not bother.

kmshin1397 replied:
@dmitryduev do we want this sorted by increasing or decreasing hjd?

dmitryduev replied:
Let's do increasing hjd, i.e. the 0th element would be the oldest data point.

dmitryduev reviewed (commented):
Looks great, thanks a bunch @kmshin1397! 
Please add the script and istarmap.py [here](https://github.com/dmitryduev/kowalski/blob/master/ingester.Dockerfile#L40), otherwise just one request/comment, and it's good to go in.

kmshin1397 commented:
@dmitryduev Sorting by hjd is now added; also updated the Dockerfile like you suggested.

I tried out the sorting on a couple of smaller matchfiles (10-15k sources) and saw something like 1-2% increase in time. I don't think that should change drastically with the full set of matchfiles, since the number of data points for each source is magnitudes smaller than the number of sources being processed. So I think we're roughly just adding some small O(n) step to the overall runtime if we think about it in terms of n=number of sources.

dmitryduev commented:
@kmshin1397 Nice, thanks! Should I merge once green?

kmshin1397 commented:
> @kmshin1397 Nice, thanks! Should I merge once green?

Yup I think this is good to merge. Would make sure it reaches green though just to make sure I actually fixed the failure in the Dockerfile.

kmshin1397 commented:
Ok @dmitryduev now green :)
